File: stacking_env/game.py
"""Core Stacking Game game logic.

Board layout:
  Level 0: 4x4 = 16 positions (indices 0-15)
  Level 1: 3x3 = 9  positions (indices 16-24)
  Level 2: 2x2 = 4  positions (indices 25-28)
  Level 3: 1x1 = 1  position  (index 29)

Cell values: 0=empty, 1=player_0 (light), 2=player_1 (dark)
"""

import numpy as np

# Level offsets and sizes
LEVEL_OFFSET = [0, 16, 25, 29]
LEVEL_SIZE = [4, 3, 2, 1]
NUM_CELLS = 30

# Precompute: for each cell on levels 1-3, which 4 cells below support it
SUPPORT_MAP: dict[int, tuple[int, int, int, int]] = {}
for level in range(1, 4):
    s = LEVEL_SIZE[level]
    s_below = LEVEL_SIZE[level - 1]
    off = LEVEL_OFFSET[level]
    off_below = LEVEL_OFFSET[level - 1]
    for r in range(s):
        for c in range(s):
            idx = off + r * s + c
            SUPPORT_MAP[idx] = (
                off_below + r * s_below + c,
                off_below + r * s_below + (c + 1),
                off_below + (r + 1) * s_below + c,
                off_below + (r + 1) * s_below + (c + 1),
            )

# Precompute: for each cell, which cells above rest on it
RESTING_ON: dict[int, list[int]] = {i: [] for i in range(NUM_CELLS)}
for upper, lowers in SUPPORT_MAP.items():
    for lo in lowers:
        RESTING_ON[lo].append(upper)

# Precompute: all 2x2 squares per level (for square detection)
SQUARES_PER_LEVEL: list[list[tuple[int, ...]]] = []
for level in range(4):
    s = LEVEL_SIZE[level]
    off = LEVEL_OFFSET[level]
    squares = []
    for r in range(s - 1):
        for c in range(s - 1):
            squares.append((
                off + r * s + c,
                off + r * s + (c + 1),
                off + (r + 1) * s + c,
                off + (r + 1) * s + (c + 1),
            ))
    SQUARES_PER_LEVEL.append(squares)


def cell_level(idx: int) -> int:
    for lv in range(3, -1, -1):
        if idx >= LEVEL_OFFSET[lv]:
            return lv
    return 0


class StackingGame:
    def __init__(self):
        self.reset()

    def reset(self):
        self.board = np.zeros(NUM_CELLS, dtype=np.int8)
        self.reserves = [15, 15]  # player 0, player 1
        self.current_player = 0  # 0=light, 1=dark
        self.phase = "place"  # "place" or "take_back"
        self.take_backs_remaining = 0
        self.done = False
        self.winner = -1  # -1=none, 0 or 1

    def _player_val(self, player: int) -> int:
        return player + 1

    def is_supported(self, idx: int) -> bool:
        """A cell is supported if it's on level 0 or all 4 cells below are occupied."""
        if idx < 16:
            return True
        return all(self.board[s] != 0 for s in SUPPORT_MAP[idx])

    def is_uncovered(self, idx: int) -> bool:
        """A sphere is uncovered if no sphere above rests on it."""
        return all(self.board[upper] == 0 for upper in RESTING_ON[idx])

    def _formed_square(self, idx: int, player_val: int) -> bool:
        """Check if placing at idx completed any 2x2 square of player_val."""
        lv = cell_level(idx)
        for sq in SQUARES_PER_LEVEL[lv]:
            if idx in sq and all(self.board[c] == player_val for c in sq):
                return True
        return False

    def legal_place_actions(self) -> list[int]:
        """Return list of legal PLACE action indices.
        Actions 0-29: place from reserve at position i.
        Actions 30+: raise from src to dst, encoded as 30 + src*30 + dst.
        """
        pv = self._player_val(self.current_player)
        actions = []

        # Place from reserve
        if self.reserves[self.current_player] > 0:
            for i in range(NUM_CELLS):
                if self.board[i] == 0 and self.is_supported(i):
                    actions.append(i)

        # Raise: move own uncovered sphere to a higher supported empty cell
        for src in range(NUM_CELLS):
            if self.board[src] != pv or not self.is_uncovered(src):
                continue
            src_level = cell_level(src)
            # Temporarily remove to check support
            self.board[src] = 0
            for dst in range(NUM_CELLS):
                if self.board[dst] != 0:
                    continue
                if cell_level(dst) <= src_level:
                    continue
                if self.is_supported(dst):
                    actions.append(30 + src * 30 + dst)
            self.board[src] = pv

        return actions

    def legal_take_back_actions(self) -> list[int]:
        """Return list of legal TAKE_BACK action indices.
        Action 0: pass (done taking back).
        Actions 1-30: take back sphere at position (action-1).
        """
        actions = [0]  # can always pass
        pv = self._player_val(self.current_player)
        for i in range(NUM_CELLS):
            if self.board[i] == pv and self.is_uncovered(i):
                actions.append(i + 1)
        return actions

    def apply_place(self, action: int):
        """Apply a PLACE action. Returns True if a same-color square was formed."""
        pv = self._player_val(self.current_player)
        if action < 30:
            # Place from reserve
            pos = action
            self.board[pos] = pv
            self.reserves[self.current_player] -= 1
        else:
            # Raise
            code = action - 30
            src, dst = divmod(code, 30)
            self.board[src] = 0
            self.board[dst] = pv
            pos = dst

        # Check win: placed at top
        if pos == 29:
            self.done = True
            self.winner = self.current_player
            self.phase = "place"
            return False

        # Check if square formed
        if self._formed_square(pos, pv):
            self.phase = "take_back"
            self.take_backs_remaining = 2
            return True

        # End turn
        self._end_turn()
        return False

    def apply_take_back(self, action: int):
        """Apply a TAKE_BACK action."""
        if action == 0:
            # Pass — done taking back
            self.take_backs_remaining = 0
        else:
            pos = action - 1
            self.board[pos] = 0
            self.reserves[self.current_player] += 1
            self.take_backs_remaining -= 1

        if self.take_backs_remaining <= 0:
            self.phase = "place"
            self._end_turn()

    def _end_turn(self):
        self.phase = "place"
        self.current_player = 1 - self.current_player
        # Check if next player can move
        if not self.legal_place_actions():
            self.done = True
            self.winner = 1 - self.current_player  # current player loses

    def step(self, action: int):
        """Execute one action. The action space depends on self.phase."""
        if self.phase == "place":
            self.apply_place(action)
        else:
            self.apply_take_back(action)

File: pylos-web/src/game.ts
// Pylos game logic — direct port from Python pylos_env/game.py

export const LEVEL_OFFSET = [0, 16, 25, 29];
export const LEVEL_SIZE = [4, 3, 2, 1];
export const NUM_CELLS = 30;

// Support map: for each cell on levels 1-3, which 4 cells below support it
export const SUPPORT_MAP: Map<number, [number, number, number, number]> = new Map();
for (let level = 1; level < 4; level++) {
  const s = LEVEL_SIZE[level], sb = LEVEL_SIZE[level - 1];
  const off = LEVEL_OFFSET[level], offb = LEVEL_OFFSET[level - 1];
  for (let r = 0; r < s; r++)
    for (let c = 0; c < s; c++)
      SUPPORT_MAP.set(off + r * s + c, [
        offb + r * sb + c, offb + r * sb + c + 1,
        offb + (r + 1) * sb + c, offb + (r + 1) * sb + c + 1,
      ]);
}

// Resting on: for each cell, which cells above rest on it
export const RESTING_ON: number[][] = Array.from({ length: NUM_CELLS }, () => []);
for (const [upper, lowers] of SUPPORT_MAP) for (const lo of lowers) RESTING_ON[lo].push(upper);

// All 2x2 squares per level
export const SQUARES_PER_LEVEL: number[][][] = [];
for (let level = 0; level < 4; level++) {
  const s = LEVEL_SIZE[level], off = LEVEL_OFFSET[level], squares: number[][] = [];
  for (let r = 0; r < s - 1; r++)
    for (let c = 0; c < s - 1; c++)
      squares.push([off + r * s + c, off + r * s + c + 1, off + (r + 1) * s + c, off + (r + 1) * s + c + 1]);
  SQUARES_PER_LEVEL.push(squares);
}

export function cellLevel(idx: number): number {
  for (let lv = 3; lv >= 0; lv--) if (idx >= LEVEL_OFFSET[lv]) return lv;
  return 0;
}

// Board positions: (x, y, z) for each cell
const SQRT2 = Math.sqrt(2);
export const BOARD_POS: [number, number, number][] = [];
{
  let idx = 0;
  for (let level = 0; level < 4; level++) {
    const s = LEVEL_SIZE[level], z = 1 + level * SQRT2, half = s - 1;
    for (let r = 0; r < s; r++)
      for (let c = 0; c < s; c++) {
        BOARD_POS[idx] = [-half + c * 2, -half + r * 2, z];
        idx++;
      }
  }
}

// Reserve positions
export function reservePositions(side: -1 | 1): [number, number, number][] {
  const yBase = side * 8, pos: [number, number, number][] = [];
  for (let i = 0; i < 15; i++) {
    const row = Math.floor(i / 8), col = i % 8;
    pos.push([-7 + col * 2, yBase + side * row * 2, 1]);
  }
  return pos;
}

export type Phase = 'place' | 'take_back';

export interface GameSnapshot {
  board: number[];
  reserves: [number, number];
  currentPlayer: number;
  phase: Phase;
  takeBacksRemaining: number;
  done: boolean;
  winner: number;
}

export class PylosGame {
  board: number[];       // 0=empty, 1=player0(light), 2=player1(dark)
  reserves: [number, number];
  currentPlayer: number; // 0 or 1
  phase: Phase;
  takeBacksRemaining: number;
  done: boolean;
  winner: number;        // -1=none

  constructor() { this.board = []; this.reserves = [15, 15]; this.currentPlayer = 0; this.phase = 'place'; this.takeBacksRemaining = 0; this.done = false; this.winner = -1; this.reset(); }

  reset() {
    this.board = new Array(NUM_CELLS).fill(0);
    this.reserves = [15, 15];
    this.currentPlayer = 0;
    this.phase = 'place';
    this.takeBacksRemaining = 0;
    this.done = false;
    this.winner = -1;
  }

  snapshot(): GameSnapshot {
    return { board: [...this.board], reserves: [...this.reserves] as [number, number], currentPlayer: this.currentPlayer, phase: this.phase, takeBacksRemaining: this.takeBacksRemaining, done: this.done, winner: this.winner };
  }

  restore(s: GameSnapshot) {
    this.board = [...s.board]; this.reserves = [...s.reserves] as [number, number];
    this.currentPlayer = s.currentPlayer; this.phase = s.phase;
    this.takeBacksRemaining = s.takeBacksRemaining; this.done = s.done; this.winner = s.winner;
  }

  playerVal(p: number) { return p + 1; }

  isSupported(idx: number): boolean {
    if (idx < 16) return true;
    const sup = SUPPORT_MAP.get(idx)!;
    return sup.every(s => this.board[s] !== 0);
  }

  isUncovered(idx: number): boolean {
    return RESTING_ON[idx].every(u => this.board[u] === 0);
  }

  private formedSquare(idx: number, pv: number): boolean {
    const lv = cellLevel(idx);
    for (const sq of SQUARES_PER_LEVEL[lv])
      if (sq.includes(idx) && sq.every(c => this.board[c] === pv)) return true;
    return false;
  }

  legalPlaceActions(): number[] {
    const pv = this.playerVal(this.currentPlayer), actions: number[] = [];
    if (this.reserves[this.currentPlayer] > 0)
      for (let i = 0; i < NUM_CELLS; i++)
        if (this.board[i] === 0 && this.isSupported(i)) actions.push(i);
    for (let src = 0; src < NUM_CELLS; src++) {
      if (this.board[src] !== pv || !this.isUncovered(src)) continue;
      const srcLv = cellLevel(src);
      this.board[src] = 0;
      for (let dst = 0; dst < NUM_CELLS; dst++)
        if (this.board[dst] === 0 && cellLevel(dst) > srcLv && this.isSupported(dst))
          actions.push(30 + src * 30 + dst);
      this.board[src] = pv;
    }
    return actions;
  }

  legalTakeBackActions(): number[] {
    const actions = [0], pv = this.playerVal(this.currentPlayer);
    for (let i = 0; i < NUM_CELLS; i++)
      if (this.board[i] === pv && this.isUncovered(i)) actions.push(i + 1);
    return actions;
  }

  applyPlace(action: number): boolean {
    const pv = this.playerVal(this.currentPlayer);
    let pos: number;
    if (action < 30) {
      pos = action; this.board[pos] = pv; this.reserves[this.currentPlayer]--;
    } else {
      const code = action - 30, src = Math.floor(code / 30), dst = code % 30;
      this.board[src] = 0; this.board[dst] = pv; pos = dst;
    }
    if (pos === 29) { this.done = true; this.winner = this.currentPlayer; return false; }
    if (this.formedSquare(pos, pv)) { this.phase = 'take_back'; this.takeBacksRemaining = 2; return true; }
    this.endTurn(); return false;
  }

  applyTakeBack(action: number) {
    if (action === 0) { this.takeBacksRemaining = 0; }
    else { const pos = action - 1; this.board[pos] = 0; this.reserves[this.currentPlayer]++; this.takeBacksRemaining--; }
    if (this.takeBacksRemaining <= 0) { this.phase = 'place'; this.endTurn(); }
  }

  private endTurn() {
    this.phase = 'place';
    this.currentPlayer = 1 - this.currentPlayer;
    if (this.legalPlaceActions().length === 0) { this.done = true; this.winner = 1 - this.currentPlayer; }
  }
}

export function describeAction(action: number, phase: Phase, player: number): string {
  const name = player === 0 ? 'Light' : 'Dark';
  if (phase === 'place') {
    if (action < 30) return `${name} places at ${action} (L${cellLevel(action)})`;
    const code = action - 30, src = Math.floor(code / 30), dst = code % 30;
    return `${name} raises ${src} to ${dst}`;
  }
  return action === 0 ? `${name} passes take-back` : `${name} takes back ${action - 1}`;
}
